### PR TITLE
[IO-856] test for background deletion

### PR DESCRIPTION
**Pull request by thhart**
test for exceptional listing of deleted files

See [IO-856](https://issues.apache.org/jira/browse/IO-856)

### CODE REVIEW

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 1]
```diff
@@ -222,4 +219,48 @@ public void testListFilesWithDeletion() throws IOException {
...
+        final int waitTime = 10000;
+        final java.util.concurrent.CompletableFuture<Void> c1 = java.util.concurrent.CompletableFuture.runAsync(() -> {
+            long endTime = System.currentTimeMillis() + waitTime;
+            while (System.currentTimeMillis() < endTime) {
+                final File file = new File(dir.getAbsolutePath(), java.util.UUID.randomUUID() + ".deletetester");
+                file.deleteOnExit();
+                try {
+                    new java.io.BufferedOutputStream(java.nio.file.Files.newOutputStream(file.toPath())).write("TEST".getBytes(java.nio.charset.StandardCharsets.UTF_8));
+                } catch (Exception e) {
+                    throw new RuntimeException("could not create test file: " + file.getAbsolutePath(), e);
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:241]: Don't throw `RuntimeException`, use JUnit's `fail()` method.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 2]
```diff
@@ -222,4 +219,48 @@ public void testListFilesWithDeletion() throws IOException {
...
+                }
+                if (!file.delete()) {
+                    throw new RuntimeException("could not delete test file: " + file.getAbsolutePath());
+                }
+            }
+        });
+
+        final java.util.concurrent.CompletableFuture<Void> c2 = java.util.concurrent.CompletableFuture.runAsync(() -> {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:249]: Use an import for `java.util.concurrent.CompletableFuture`, don't use FQCNs.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 3]
```diff
@@ -222,4 +219,48 @@ public void testListFilesWithDeletion() throws IOException {
...
+            }
+        }
+        final int waitTime = 10000;
+        final java.util.concurrent.CompletableFuture<Void> c1 = java.util.concurrent.CompletableFuture.runAsync(() -> {
+            long endTime = System.currentTimeMillis() + waitTime;
+            while (System.currentTimeMillis() < endTime) {
+                final File file = new File(dir.getAbsolutePath(), java.util.UUID.randomUUID() + ".deletetester");
+                file.deleteOnExit();
+                try {
+                    new java.io.BufferedOutputStream(java.nio.file.Files.newOutputStream(file.toPath())).write("TEST".getBytes(java.nio.charset.StandardCharsets.UTF_8));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:239]: Use imports.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 4]
```diff
@@ -222,4 +219,48 @@ public void testListFilesWithDeletion() throws IOException {
...
+                }
+            }
+        });
+
+        final java.util.concurrent.CompletableFuture<Void> c2 = java.util.concurrent.CompletableFuture.runAsync(() -> {
+            long endTime = System.currentTimeMillis() + waitTime;
+            try {
+                while (System.currentTimeMillis() < endTime) {
+                    final Collection<File> files = FileUtils.listFiles(dir, new String[]{"\\.deletetester"}, false);
+                    files.size();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:254]: What does calling `size()` do here? Should there be an assert for the result? is the call needed at all?

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hello @thhart 

Thank you for your PR.

Please see my comments.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 5]
```diff
@@ -16,27 +16,26 @@
...
 import java.util.stream.Stream;
 
-import org.apache.commons.io.filefilter.FileFilterUtils;
-import org.apache.commons.io.filefilter.IOFileFilter;
-import org.apache.commons.io.function.Uncheck;
-import org.apache.commons.lang3.function.Consumers;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.io.TempDir;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:38]: Do not use star imports.

Please run `mvn` by itself before you push. 

In this case, you would have discovered this error:
```
(imports) AvoidStarImport: Using the '.*' form of import should be avoided 
```

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 6]
```diff
@@ -222,4 +221,47 @@ public void testListFilesWithDeletion() throws IOException {
...
     }
 
+    @Test
+    public void testListFilesWithDeletionThreaded() throws java.util.concurrent.ExecutionException, InterruptedException {
+        // test for IO-856
+        // create random directory in tmp, create the directory if it does not exist
+        final File dir = FileUtils.getTempDirectory();
+        if (!dir.exists()) {
+            if (!dir.mkdirs()) {
+                throw new RuntimeException("could not create image file path: " + dir.getAbsolutePath());
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:231]: As noted previously, use JUnit's `fail()` method.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@thhart 

Thank you for your update.

You did not apply my comments throughout. See my additional comments.

TY

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 7]
```diff
@@ -222,4 +226,47 @@ public void testListFilesWithDeletion() throws IOException {
         assertEquals(4, list.size());
     }
 
+    @Test
+    public void testListFilesWithDeletionThreaded() throws java.util.concurrent.ExecutionException, InterruptedException {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:230]: You don't need a fully qualified class name here.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 8]
```diff
@@ -222,4 +226,47 @@ public void testListFilesWithDeletion() throws IOException {
...
+        // test for IO-856
+        // create random directory in tmp, create the directory if it does not exist
+        final File dir = FileUtils.getTempDirectory();
+        if (!dir.exists()) {
+            if (!dir.mkdirs()) {
+                fail("could not create image file path: " + dir.getAbsolutePath());
+            }
+        }
+        final int waitTime = 10000;
+        final CompletableFuture<Void> c1 = java.util.concurrent.CompletableFuture.runAsync(() -> {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:240]: You don't need a fully qualified class name here.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 9]
```diff
@@ -222,4 +226,47 @@ public void testListFilesWithDeletion() throws IOException {
...
+            long endTime = System.currentTimeMillis() + waitTime;
+            while (System.currentTimeMillis() < endTime) {
+                final File file = new File(dir.getAbsolutePath(), java.util.UUID.randomUUID() + ".deletetester");
+                file.deleteOnExit();
+                try {
+                    new BufferedOutputStream(java.nio.file.Files.newOutputStream(file.toPath())).write("TEST".getBytes(java.nio.charset.StandardCharsets.UTF_8));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:246]: You don't need a fully qualified class name here.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 10]
```diff
@@ -222,4 +221,47 @@ public void testListFilesWithDeletion() throws IOException {
...
+                } catch (Exception e) {
+                    fail("could not create test file: " + file.getAbsolutePath(), e);
+                }
+                if (!file.delete()) {
+                    fail("could not delete test file: " + file.getAbsolutePath());
+                }
+            }
+        });
+
+        final CompletableFuture<Void> c2 = java.util.concurrent.CompletableFuture.runAsync(() -> {
```
[the remark this answers is not in the document]
Comment (garydgregory) [reply src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:256]: You don't need a fully qualified class name here.

File: src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java [hunk 11]
```diff
@@ -211,7 +214,8 @@ public void testListFilesWithDeletion() throws IOException {
         final String[] extensions = {"xml", "txt"};
         final List<File> list;
         final File xFile = new File(temporaryFolder, "x.xml");
-        xFile.createNewFile();
+        if(!xFile.createNewFile())
+            fail("could not create test file: " + xFile);
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsListFilesTest.java:218]: Always use a block.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@thhart 

Please see my comments, same as before ;-)

**Comment by garydgregory** [comment]
Hello @thhart 

Thank you for your update.

Interesting that the build on Java 24-ea is green.

Ubuntu shows:
```
FileUtilsListFilesTest.setUp:77 » FileSystem /tmp/junit-6966172520746092400/dummy-build.xml: Too many open files
```
Is that the point of the test?

Another builds show:
```
[INFO] There are 5 errors reported by Checkstyle 9.3 with D:\a\commons-io\commons-io/src/conf/checkstyle.xml ruleset.
Error:  src\test\java\org\apache\commons\io\FileUtilsListFilesTest.java:[27,1] (imports) ImportOrder: Wrong order for 'java.io.BufferedOutputStream' import.
Error:  src\test\java\org\apache\commons\io\FileUtilsListFilesTest.java:[41,1] (imports) ImportOrder: Wrong order for 'org.junit.jupiter.api.Assertions.assertEquals' import.
Error:  src\test\java\org\apache\commons\io\FileUtilsListFilesTest.java:[220,9] (whitespace) WhitespaceAfter: 'if' is not followed by whitespace.
Error:  src\test\java\org\apache\commons\io\FileUtilsListFilesTest.java:[245,18] (coding) FinalLocalVariable: Variable 'endTime' should be declared final.
Error:  src\test\java\org\apache\commons\io\FileUtilsListFilesTest.java:[261,18] (coding) FinalLocalVariable: Variable 'endTime' should be declared final.
```
Please run `mvn` by itself to run all build checks and fix errors before you push.

**Comment by thhart** [comment]
> FileUtilsListFilesTest.setUp:77 » FileSystem /tmp/junit-6966172520746092400/dummy-build.xml: Too many open files
> ```
> 
> Is that the point of the test?

No, this is not failing in my environment. Maybe your Ubuntu is too restrictive in this.


> [INFO] There are 5 errors reported by Checkstyle 9.3 with D:\a\commons-io\commons-io/src/conf/checkstyle.xml ruleset.

This is changed now.

> Please run `mvn` by itself to run all build checks and fix errors before you push.

```
[ERROR] Errors: 
[ERROR]   FileUtilsListFilesTest.testListFilesWithDeletionThreaded:273 » Execution org.opentest4j.AssertionFailedError: this should not happen
[INFO] 
[ERROR] Tests run: 4246, Failures: 0, Errors: 1, Skipped: 24
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  05:18 min
[INFO] Finished at: 2024-11-08T09:09:46+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.5.1:test (default-test) on project commons-io: 
[ERROR] 
[ERROR] Please refer to /home/th/dev/os/commons-io/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.

```
This fails in m Ubuntu system and in other Linux flavour. Can not say for other systems.

**Comment by garydgregory** [comment]
@thhart 

Thank you for your updates. I'll take a look over the next couple of days.

**Comment by garydgregory** [comment]
Interesting: This is not reproducible on Windows in this CI for Java 8, 11, 17, 21, and 23.

**Comment by garydgregory** [comment]
Interesting meta-failure on Ubuntu:
```
[INFO] Running org.apache.commons.io.FileUtilsListFilesTest
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message can't create byte arrau at JPLISAgent.c line: 813
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message can't create byte arrau at JPLISAgent.c line: 813
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message can't create byte arrau at JPLISAgent.c line: 813
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message can't create byte arrau at JPLISAgent.c line: 813
```

**Comment by garydgregory** [comment]
Hello @thhart 

- I added a slightly modified version of this test in `FileUtilsListFilesTest.testListFilesWithDeletionThreaded()`
- This test is only enabled on Windows
- There are hidden failures on GitHub CI on Ubuntu and macOS

**Comment by garydgregory** [comment]
@thhart 

Please test git master. Closing this PR in favor or #699.

**Comment by thhart** [comment]
Looks ok now, have no problems any more.
(There was a wrong escape in the extension, it is not a pattern. I created a small patch for it.)

**Pull request closed without merging**